### PR TITLE
CAFS streaming

### PR DESCRIPTION
**Pull request by kerneltime**
This supersedes https://github.com/oneconcern/datamon/pull/177
Includes 
1. Caching of working set of Blobs
2. Fixes for passing CI.

### CODE REVIEW

File: pkg/cafs/reader.go [hunk 1]
```diff
@@ -159,6 +170,85 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
 
 }
 
+func (r *chunkReader) calculateKeyAndOffset(off int64) (index int64, offset int64) {
+	index = off / int64(r.leafSize)
+	offset = off - (index * int64(r.leafSize))
```
Comment (ransomw1c) [inline pkg/cafs/reader.go]: `off % int64(r.leafSize)` instead?

File: pkg/cafs/reader.go [hunk 2]
```diff
@@ -159,6 +170,85 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
...
+				r.lru.Add(k.StringWithPrefix(r.prefix), b)
+			}(nextIndex)
+		}
+	}
+
+	for {
+		// Fetch Blob
+		var buffer []byte
+		key := r.keys[index]
+		if r.lru.Contains(key.StringWithPrefix(r.prefix)) {
```
Comment (ransomw1c) [inline pkg/cafs/reader.go]: on the first pass of the outer `for` loop, it's possible that this `if` block will need to account for `offset`

File: pkg/cafs/reader.go [hunk 3]
```diff
@@ -159,6 +170,85 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
...
+
+	// Calculate first key and offset.
+	index, offset := r.calculateKeyAndOffset(off)
+	if index >= int64(len(r.keys)) {
+		return 0, nil
+	}
+
+	var read int
+
+	seekHead := func() {
```
Comment (ransomw1c) [inline pkg/cafs/reader.go]: might this broken out as a separate function and/or could there be a comment regarding its intent?

in particular, i'm curious to hear more about the (presumed) analogy between what this function is doing and [`seek`](http://man.cat-v.org/plan_9/2/seek) calls on file-handles.

File: pkg/cafs/reader.go (hunk 1 above)
Comment (kerneltime) [reply pkg/cafs/reader.go]: that works too

File: pkg/cafs/reader.go (hunk 3 above)
Comment (kerneltime) [reply pkg/cafs/reader.go]: seek ahead implies that the reader is seeking ahead to the next blob in the sequence.
The goal here is to serve reads from memory as far as possible when streaming a file. This way a read request never blocks on read completing from backend. Will add the comments.

**Comment by ransomw1c** [comment]
the overall idea i'm taking from this change is that it's an implementation of streaming from `gcs` thru `cafs` to the bundle...

... the in-memory cache is added to the stream at the `cafs` level in order to prevent "stutters" on streaming.

**Comment by kerneltime** [comment]
> the overall idea i'm taking from this change is that it's an implementation of streaming from `gcs` thru `cafs` to the bundle...
> 
> ... the in-memory cache is added to the stream at the `cafs` level in order to prevent "stutters" on streaming.

Yes

**Pull request merged**